### ISSUE TITLE
FR: Generalizable and reusable "rebase behavior"

### ISSUE THREAD

**Issue by arxanas** [martinvonz/jj#4708]
**Is your feature request related to a problem? Please describe.**

There's been a proliferation of flags added to various commands that create or rebase commits. For example, `--insert-before`/`--insert-after` should be added to many commands, but hasn't been yet. Rebasing is a core competency and reason to use jj. I think we should invest in a unified flexible design to control "rebase behavior" rather than add more flags.

Most commands that create or rebase commits could reuse some of these options, including:

- `jj abandon`
- `jj commit`
- `jj describe`
- `jj duplicate`
- `jj new`
- `jj parallelize`
- `jj rebase`
- `jj split`
- `jj squash`

**Describe the solution you'd like**

Roughly speaking `jj rebase --dest X` (and similar for other commands) should accept an expression representing "rebase behavior" (and also accept a revset for backward compatibility and the common case).

As an example, suppose you're using a "mega-merge" strategy. You have a bunch of commits in `X` that you want to insert before the mega-merge commit as standalone commits. You want to take this opportunity to also write some helpful messages for each commit. If this wouldn't merge cleanly, then you want to abort the operation rather than produce conflicts. An example invocation might look like this:

```console
$ jj rebase -r X -d 'behavior(after=trunk(), before=@, topology="horizontal", empty="keep", message="prompt", merge="abort")'
```

However, if you have a commit in `X` and you need to split it as part of doing the mega-merge, then maybe you'll issue an invocation like this:

```console
$ jj split -r X -d 'behavior(after=trunk(), before=@, topology="horizontal", message="prompt", merge="ours")'
```

To demonstrate the general applicability of this design, see the issues linked below.

Given a commit set `X` that's being rebased, the following are some relevant axes to control rebase behavior:
- **Location**: one or more of the following:
  - "after" revset `Y`: commit set `X` is rebased to be children of `Y`. This is the current behavior of `jj rebase -r X -d Y`.
  - "before" revset `Z`: after rebasing the commit set, the commits in `Z` will be rebased as children of it. This is the current behavior of `jj rebase -r X --insert-before Z`.
  - See: https://github.com/martinvonz/jj/issues/1188, https://github.com/martinvonz/jj/issues/1531, https://github.com/martinvonz/jj/issues/3518, https://github.com/martinvonz/jj/issues/3772, https://github.com/martinvonz/jj/issues/4468
- **Topology**: exactly one of the following:
  - "preserve": Default. If a commit in `X` was an ancestor of another commit in `X` before the rebase, then it remains so after the rebase. This is the current behavior of `jj rebase`.
  - "horizontal": Commits in `X` are rebased individually to be siblings. This is the behavior of `jj parallelize`.
  - "vertical": Commits in `X` are rebased to be a chain of children (in an arbitrary order). No current equivalent.
  - See: https://github.com/martinvonz/jj/issues/1079, https://github.com/martinvonz/jj/issues/2274, https://github.com/martinvonz/jj/issues/2600, https://github.com/martinvonz/jj/issues/4351
- **Signature**: for each combination of author/committer + name/email/timestamp:
  - "preserve": Default. Preserve the existing signature value.
  - "set" to `Y`: Set the signature value to `Y`. Some current equivalents.
  - "reset": Overwrite the signature value to the current user's signature value. Some current equivalents.
  - Potentially something about GPG-signing?
  - See: https://github.com/martinvonz/jj/issues/58, https://github.com/martinvonz/jj/issues/328, https://github.com/martinvonz/jj/issues/2000,  https://github.com/martinvonz/jj/issues/3008, https://github.com/martinvonz/jj/issues/3565, https://github.com/martinvonz/jj/issues/4170
- **Message**:
  - "preserve": Default. Preserve the existing message.
  - "set" to `Y`: For each commit in `X`, set the message to string (or template) `Y`.
  - "prompt": Interactively prompt to edit all messages for commits in `X`.
  - related: "add/set footer": Add or set a Git footer for the commit.
  - See https://github.com/martinvonz/jj/issues/1155, https://github.com/martinvonz/jj/issues/2669, https://github.com/martinvonz/jj/issues/3060
- **Empty commits**:
  - "drop": Default. Drop any commits in `X` that are empty.
  - "preserve": Preserve all commits in `X`, even those that are empty.
  - "keep": Drop commits in `X` that were non-empty before the rebase but empty after the rebase. (Should this be the new default?) Related to `--keep-emptied`.
  - related: "deduplicate": Determine if any commits have already been applied via patch ID semantics.
  - See: https://github.com/martinvonz/jj/issues/229, https://github.com/martinvonz/jj/issues/1039, https://github.com/martinvonz/jj/issues/1399, https://github.com/martinvonz/jj/issues/1406, https://github.com/martinvonz/jj/issues/1946, https://github.com/martinvonz/jj/issues/2979, https://github.com/martinvonz/jj/issues/3120, https://github.com/martinvonz/jj/issues/3766, https://github.com/martinvonz/jj/issues/4641, https://github.com/martinvonz/jj/issues/4668
- **Change ID**:
  - "preserve": Default. Preserve the existing change ID.
  - "set" to `Y`: Set the change ID to `Y` explicitly. No current equivalent.
  - "reset": Overwrite the change ID with a freshly-generated one. No current equivalent.
  - See: https://github.com/martinvonz/jj/issues/490, https://github.com/martinvonz/jj/issues/4451
- **Conflict resolution**:
  - "preserve": Default. Allow conflicts to be generated and recorded.
  - "reparent": Set the parents of commits in `X` without actually doing "rebasing" (patch application).
  - "abort": If conflicts would be generated, instead abort the operation.
  - "ours"/"theirs": When conflicts are generated, use one of the sides to resolve them. (There are several additional aspects that can be configured here.)
  - See: https://github.com/martinvonz/jj/issues/47, https://github.com/martinvonz/jj/issues/555, https://github.com/martinvonz/jj/issues/1027, https://github.com/martinvonz/jj/issues/3386, https://github.com/martinvonz/jj/issues/3540, https://github.com/martinvonz/jj/issues/2849, https://github.com/martinvonz/jj/issues/3805, https://github.com/martinvonz/jj/issues/4152
- **Bookmark moving**:
  - Not sure about this one, but there might be generally-relevant configuration. If we adopt topics instead of bookmarks, then it might also make sense to configure topic behavior here.
  - See: https://github.com/martinvonz/jj/issues/2338, https://github.com/martinvonz/jj/issues/3419, https://github.com/martinvonz/jj/issues/3505, https://github.com/martinvonz/jj/issues/3805

**Describe alternatives you've considered**

- Preserve current behavior. Add flags to each command that might need them. I think it's extra work overall and results in a less flexible/coherent system. However, command-line flags are generally more discoverable/accessible than the revset language.

**Additional context**

This also covers some use-cases for https://github.com/martinvonz/jj/issues/3814, but is not a complete substitute.

**Comment by joyously** [comment]
This seems like a lot of options. Hopefully, all will be considered in the context of `undo` and of simultaneous access (scripts in two windows).

**Comment by arxanas** [comment]
> This seems like a lot of options. Hopefully, all will be considered in the context of undo and of simultaneous access (scripts in two windows).

I don't think we need to do anything special to support undo for this design. I believe we basically snapshot the repo before and after the operation, without regard to which operations actually ran. Simultaneous modifications are a problem, but not specific to this either.
